### PR TITLE
implemented RegistrationServiceImplTest from RegistrationServiceImpl

### PR DESCRIPTION
**Pull request by vsych321**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -1,14 +1,41 @@
...
+    }
+
+    private static void checkLoginValidation(User user) {
+        if (user.getLogin() == null || user.getLogin().length() < VALID_LOGIN) {
+            throw new CustomException("Error! Login must have at least length " + VALID_LOGIN);
+        }
+    }
+
+    private void checkPasswordValidation(User user) {
+        if (user.getPassword() == null || user.getPassword().length() < VALID_PASSWORD) {
```
Comment (Cayman4010) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:29]: it`s the same

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -1,14 +1,41 @@
...
+            throw new CustomException("Error! Password must have at least length "
+                    + VALID_PASSWORD);
+        }
+
+    }
+
+    private static void checkAgeValidation(User user) {
+        if (user.getAge() == null || user.getAge() < VALID_AGE) {
```
Comment (Cayman4010) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:37]: and here

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    @Test
+    void register_UserIsValid_Ok() {
+        user = new User("Arsenal", "Ar789al", 20);
+        assertEquals(user, registrationService.register(user));
+    }
+
+    @Test
+    void register_UserNotValid_notOk() {
+        user = new User("daff", "gadaz",11);
```
Comment (Cayman4010) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:30]: comma before 11

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    void setUp() {
+        registrationService = new RegistrationServiceImpl();
+    }
+
+    @Test
+    void register_UserIsValid_Ok() {
+        user = new User("Arsenal", "Ar789al", 20);
+        assertEquals(user, registrationService.register(user));
+    }
+
```
Comment (Cayman4010) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:27]: I didn`t found the test with an adding existing user.

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    void register_NullLogin_TrowsException() {
+        user = new User(null, "validity", 21);
+        assertThrows(CustomException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_NullPassword_ThrowsException() {
```
Comment (Cayman4010) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:41]: Where NotOk in the end?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        user = new User("Arsenal", null, 21);
+        assertThrows(CustomException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_NullAge_ThrowsException() {
```
Comment (Cayman4010) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:47]: and here

**Review by Cayman4010 — CHANGES_REQUESTED** [review]
Good solution! Let`s improve it!

File: src/main/java/core/basesyntax/customexception/CustomException.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.customexception;
+
+public class CustomException extends RuntimeException {
+    public CustomException(String message) {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/customexception/CustomException.java:4]: Bad naming for exception

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 8]
```diff
@@ -1,14 +1,49 @@
 package core.basesyntax.service;
 
+import core.basesyntax.customexception.CustomException;
 import core.basesyntax.dao.StorageDao;
 import core.basesyntax.dao.StorageDaoImpl;
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int VALID_LOGIN = 6;
+    private static final int VALID_PASSWORD = 6;
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:10]: you can use only one constant MINIMAL_LENGTH_OF_CREDENTIAL_VALUE

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 9]
```diff
@@ -1,14 +1,49 @@
...
     @Override
     public User register(User user) {
-        return null;
+        checkLoginValidation(user);
+        checkPasswordValidation(user);
+        checkAgeValidation(user);
+        return storageDao.add(user);
+    }
+
+    private static void checkLoginValidation(User user) {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:22]: Why static?

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 10]
```diff
@@ -1,14 +1,49 @@
...
+    private static final int VALID_PASSWORD = 6;
+    private static final int VALID_AGE = 18;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        checkLoginValidation(user);
+        checkPasswordValidation(user);
+        checkAgeValidation(user);
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:18]: I don’t  see a check when a user with some login already exists

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 11]
```diff
@@ -1,14 +1,49 @@
...
+        return storageDao.add(user);
+    }
+
+    private static void checkLoginValidation(User user) {
+        if (user.getLogin() == null) {
+            throw new CustomException("Error! Login can't be null");
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:24]:
```suggestion
            throw new CustomException("Login can't be null");
```
It isn't an error) Let’s remove it everywhere

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private RegistrationService registrationService;
+    private User user;
+
+    @BeforeEach
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:17]: use BeforeAll

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+import core.basesyntax.customexception.CustomException;
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private RegistrationService registrationService;
+    private User user;
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:15]: make it a local variable

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Test
+    void register_UserIsValid_Ok() {
+        user = new User("Arsenal", "Ar789al", 20);
+        assertEquals(user, registrationService.register(user));
+    }
+
+    @Test
+    void register_UserNotValid_notOk() {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:29]: Which is parameter invalid? You should test only one invalid parameter in each test case. Let’s remove this case

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Test
+    void register_NullAge_ThrowsException() {
+        user = new User("Arsenal", "Ar789al", 11);
+        assertThrows(CustomException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void user_EmptyLogin_NotOk() {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:53]: the first part is a method that you test

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    void user_NotEnoughPasswordLength_NotOk() {
+        user = new User("Arsenal", "null", 21);
+        assertThrows(CustomException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void user_OverPasswordLength_Ok() {
+        user = new User("Arsenal", "Ar789al789", 21);
+        assertEquals(user, registrationService.register(user));
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:73]: you already had tested user with valid data

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  fix a few mistakes  ;)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good! Let’s fix a few mistakes ;)

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 17]
```diff
@@ -3,12 +3,56 @@
...
+                    + MINIMAL_LENGTH_OF_CREDENTIAL_VALUE);
+        }
+    }
+
+    private void checkPasswordValidation(User user) {
+        if (user.getPassword() == null) {
+            throw new RegistrationException("Password can't be null");
+        }
+        if (user.getPassword().length() < MINIMAL_LENGTH_OF_CREDENTIAL_VALUE) {
+            throw new RegistrationException("Login must have at least length "
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:44]: Don't just copy past the exception message.
Login must have at least length - not correct for check Password. 
The same for Age

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Test
+    void register_NullAge_NotOK() {
+        User user = new User("Arsenal", "Ar789al", 0);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void user_NotEnoughLoginLength_NotOk() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:53]: be consistent for test method naming. 
register_LoginLengthLessThenMinLimit_NotOk 

just an example

Check all places

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    void user_OverAge_Ok() {
+        User user = new User("Arsenal", "Ar789al", 25);
+        assertEquals(user, registrationService.register(user));
+    }
+
+    @AfterEach
+    void tearDown() {
+        Storage.PEOPLE.clear();
+    }
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:91]: Add test then age = null

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+    @Test
+    void register_UserIsValid_Ok() {
+        User user = new User("Arsenal", "Ar789al", 18);
+        assertEquals(user, registrationService.register(user));
+    }
+
+    @Test
+    void register_ExistingUser_NotOk() {
```
Comment (Andriy-dz) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:28]:
```suggestion
    void register_existingUser_notOk() {
```

**Review by Andriy-dz — APPROVED** [review]

**Review by Andriy-dz — APPROVED** [review]
a

**Review by Andriy-dz — APPROVED** [review]
a

**Review by Andriy-dz — APPROVED** [review]
a
